Drivers/Src/Display.c

```diff
@@ -1,69 +1,190 @@
 }
 
 /*
- * Initialize the Wonder Gecko communications.
+ * Initialize the Nextion display
  */
 void Display_Init() {
-    BSP_UART_Init(UART_2);
+    BSP_UART_Init(UART_3);
+    // The display sends 0x88 when ready, but that might be
+    // before we initialize our UART
 }
 
-/*
- * Updates the Gecko display with the data provided.
- * @param status the new display data
- * @return void
+/**
+ * Set the displayed velocity to vel 
  */
-void Display_SetData(display_data_t *status) {
-
-    // The update gecko will expect ten bytes: a float a some flags
-    // Each byte will be encoded as two hex digits
-    // The float will be sent in little endian format, and the flags
-    // will be encoded in the order they're declared in the struct,
-    // starting at bit 3 and going til bit 0 of the fifth byte
-    uint8_t msg[GECKO_DATA_LENGTH];
-    // Pack the flags into a byte
-    int flags = (status->cruiseEnabled << 3) |
-        (status->cruiseSet << 2) |
-        (status->regenEnabled << 1) |
-        status->canError;
-
-    #ifdef DEBUG
-    assert(sizeof(float) == 4); // Just a sanity check
-    #endif
-
-    static union fi {float f; int i;} data; // So we can interpret the float as a bitvector
-    data.f = status->speed;
-    for (int j=0; j<4; j++) {
-        encodeByte(data.i & 0xFF, msg+j*2); // enocde each byte of the float as two hex digits
-        data.i >>= 8; // Get rid of the bit we just encoded and prepare the next one
+ErrorStatus Display_SetVelocity(float vel) {
+    int32_t vel_fix = (uint32_t) floorf(vel * 10.0f);
+    return updateValue(VELOCITY, VALUE, NULL, vel_fix);
+}
+
+/**
+ * Update the cruise enable light on the display based on the given state
+ */
+ErrorStatus Display_CruiseEnable(State on) {
+    if (on == ON) {
+        return updateValue(CRUISE_ENABLE, PCO, NULL, NEXTION_GREEN);
+    } else {
+        return updateValue(CRUISE_ENABLE, PCO, NULL, NEXTION_LIGHT_GREY);
     }
+}
+
+/**
+ * Update the cruise set light on the display based on the given state
+ */
+ErrorStatus Display_CruiseSet(State on) {
+    if (on == ON) {
+        return updateValue(CRUISE_SET, PCO, NULL, NEXTION_GREEN);
+    } else {
+        return updateValue(CRUISE_SET, PCO, NULL, NEXTION_LIGHT_GREY);
+    }
+}
 
-    // Say, for example, that the speed is 38.0
-    // In IEEE-754 floating point format, it is represented as 0x42180000
-    // At this point in the code, the msg array will have "00001842" as its first eight bytes
-    
-    encodeByte((uint8_t) flags, msg+4*2);
+/**
+ * Set ERROR<idx> to err
+ * If err is an empty string, then the error will be cleared
+ */
+ErrorStatus Display_SetError(int idx, char *err) {
+    if (idx < 0 || idx > 5) return ERROR; // Index out of bounds
+    ErrorStatus err1 = updateValue(ERROR0 + idx, TEXT, err, 0);
+    ErrorStatus err2 = updateValue(ERROR0 + idx, PCO, NULL, NEXTION_GREEN);
+    return err1 && err2; // If either is error, then had an error
```

Comment:
I'm not a huge fan of relying on the `ERROR` enum to evaluate to `false`, since someone could change this in the future and not realize they need to update this code